### PR TITLE
feat(file-uploader): implements file uploader

### PR DESCRIPTION
**Pull request by chasestarr**
WIP still, but wanted to get an initial pr up here. Based quite a lot on jerome's work. Still some unknowns concerning various ui states. Only basic render has been provided.

### CODE REVIEW

File: src/file-uploader/styled-components.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    borderWidth: props.$theme.sizing.scale0,
+    boxSizing: 'border-box',
+    paddingTop: props.$theme.sizing.scale900,
+    paddingRight: props.$theme.sizing.scale800,
+    paddingBottom: props.$theme.sizing.scale900,
+    paddingLeft: props.$theme.sizing.scale800,
+  };
+});
+
+export const FileSelectButton = styled('button', (props: StylePropsT) => {
```
Comment (gergelyke) [inline src/file-uploader/styled-components.js]: can we use here the Button component?

File: src/file-uploader/README.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+* `$disabled: boolean`
+* `$multi: boolean`
+* `$isFocused: boolean`
+* `$error: boolean`
+* `$theme: theme`
+
+## Dependencies
+
+Does this component depend on any 3rd party packages or other internal components?
```
Comment (gergelyke) [inline src/file-uploader/README.md]: should we add `attr-accept` here?

File: package.json [hunk 3]
```diff
@@ -125,6 +125,7 @@
     "webpack-cli": "^3.1.0"
   },
   "dependencies": {
+    "attr-accept": "^1.1.3",
```
Comment (gergelyke) [inline package.json]: I wonder if we should use https://www.npmjs.com/package/accepts? seems like more used / maintained

**Comment by chasestarr** [comment]
it may be worth holding off on dev here until headless dropzone is released https://github.com/react-dropzone/react-dropzone/pull/717

that would streamline implementation to only include presentation

File: src/file-uploader/file-uploader.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      if (this.props.onChange) {
+        this.props.onChange({...nextFiles, event});
+      }
+
+      this.setState({
+        acceptedFiles: [
+          ...this.state.acceptedFiles,
+          ...nextFiles.acceptedFiles,
+        ],
+        rejectedFiles: [
```
Comment (nadiia) [inline src/file-uploader/file-uploader.js]: Why do we need to maintain the list of rejected files?

File: src/file-uploader/file-uploader.js [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            onDragEnter={this.onDragEnter}
+            onDragLeave={this.onDragLeave}
+            onDragOver={this.onDragOver}
+            $isDragActive={this.state.isDragActive}
+          >
+            <Block font="font450">Drop files here to upload</Block>
+            <Block font="font450" color="mono600">
+              or
+            </Block>
+            <FileSelectButton onClick={this.onClick}>
```
Comment (nadiia) [inline src/file-uploader/file-uploader.js]: How about making it a label to the hidden input?

File: src/file-uploader/file-uploader.js [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+  render() {
+    return (
+      <Root>
+        <ContentContainer>
+          <FileDragAndDrop
+            onDrop={this.onDrop}
+            onDragEnter={this.onDragEnter}
+            onDragLeave={this.onDragLeave}
+            onDragOver={this.onDragOver}
+            $isDragActive={this.state.isDragActive}
```
Comment (nadiia) [inline src/file-uploader/file-uploader.js]: it should be passed to all elements

File: src/file-uploader/file-uploader.js [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+  onWindowDragOverOrDrop = (event: SyntheticDragEvent<HTMLElement>) => {
+    event.preventDefault();
+  };
+
+  render() {
+    return (
+      <Root>
+        <ContentContainer>
+          <FileDragAndDrop
+            onDrop={this.onDrop}
```
Comment (nadiia) [inline src/file-uploader/file-uploader.js]: Would it make sense to make the whole area droppable? So the onDrag handlers will be on a container component.

File: package.json (hunk 3 above)
Comment (chasestarr) [reply package.json]: i chose this package to be consistent with `react-dropzone` https://github.com/react-dropzone/react-dropzone/blob/master/package.json#L84 which, according to a reviewer on jerome's pr, is commonly used across the FE org

File: src/file-uploader/file-uploader.js (hunk 4 above)
Comment (chasestarr) [reply src/file-uploader/file-uploader.js]: this component simply wraps the `react-dropzone` package, so re-using that api

File: src/file-uploader/file-uploader.js (hunk 7 above)
Comment (chasestarr) [reply src/file-uploader/file-uploader.js]: i've changed the markup here, so not sure if the comment still applies. but i wanted to make an outer wrapper that contains all of the elements, and just the area with the dashed border accepts drops. i'd like to display this in the design meeting, so will accept feedback on how it should operate

File: src/file-uploader/file-uploader.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+          ...styleProps
+        } = renderProps;
+
+        const prefixedStyledProps = prependStyleProps({
+          ...styleProps,
+          isDisabled: props.disabled,
+        });
+
+        return (
+          <Root>
```
Comment (nadiia) [inline src/file-uploader/file-uploader.js]: Should pass `prefixedStyledProps ` here as well

File: src/file-uploader/styled-components.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+export const StyledFileDragAndDrop = styled('div', (props: StylePropsT) => {
+  return {
+    alignItems: 'center',
+    backgroundColor: props.$theme.colors.fileUploaderBackgroundColor,
+    borderColor: props.$isDragActive
+      ? props.$theme.colors.fileUploaderBorderColorActive
+      : props.$theme.colors.fileUploaderBorderColorDefault,
+
+    borderStyle: 'dashed',
+    borderRadius: props.$theme.borders.radius200,
```
Comment (nadiia) [inline src/file-uploader/styled-components.js]: Not sure if we need to have the `$theme.borders.useRoundedCorners` check here. Something to consider.

File: src/file-uploader/styled-components.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    paddingRight: props.$theme.sizing.scale800,
+    paddingBottom: props.$theme.sizing.scale900,
+    paddingLeft: props.$theme.sizing.scale800,
+    width: '100%',
+  };
+});
+StyledFileDragAndDrop.displayName = 'StyledFileDragAndDrop';
+
+export const StyledContentMessage = styled('div', (props: StylePropsT) => ({
+  ...props.$theme.typography.font450,
```
Comment (nadiia) [inline src/file-uploader/styled-components.js]: If we set the font on the `StyledRoot` element then the inner divs should just inherit it.

**Comment by tajo** [comment]
Nice job! Is this expected for a disabled state?

disabled_outline

**Comment by chasestarr** [comment]
the inline styles are applied by the `react-dropzone` package

**Pull request merged**